File: day-3/part-2.py
def get_priority(char):
    val = ord(char)
    if val > 96:
        return val - 96
    return val - 65 + 27


def main():
    with open('day-3/input.txt', 'r') as f:
        total = 0

        lines = f.read().split('\n')
        for i in range(0, len(lines), 3):
            letters1 = {}
            letters2 = {}

            for c in lines[i]:
                letters1[c] = True

            for c in lines[i + 1]:
                letters2[c] = True

            for c in lines[i + 2]:
                if c in letters1 and c in letters2:
                    # print(c, get_priority(c))
                    total += get_priority(c)
                    break

        return total


print(main())

File: day-3/part-2.rs
use std::fs::read_to_string;
use std::collections::HashMap;

fn get_priority(c: char) -> u32 {
    let val = c as u32;
    if val > 'a' as u32 {
        return val - 'a' as u32 + 1;
    }
    return val - 'A' as u32 + 27;
}

fn get_group_val(line1: &str, line2: &str, line3: &str) -> u32 {
    let mut letters1: HashMap<char, bool> = HashMap::with_capacity(52);
    let mut letters2: HashMap<char, bool> = HashMap::with_capacity(52);
    
    for c in line1.chars() { letters1.insert(c, true); }
    for c in line2.chars() { letters2.insert(c, true); }

    for c in line3.chars() {
        if let Some(_) = letters1.get(&c) {
            if let Some(_) = letters2.get(&c) {
                // println!("{} {}", c, get_priority(c));
                return get_priority(c);
            }
        }
    }

    return 0;
}

fn main() -> std::io::Result<()> {
    let binding = read_to_string("day-3/input.txt").unwrap();
    let mut lines = binding.lines();

    let mut total: u32 = 0;
    while let Some(line) = lines.next() {
        total += get_group_val(line, lines.next().unwrap(), lines.next().unwrap());
    }

    println!("{}", total);
    Ok(())
}
